File: XamarinStore.iOS/Views/BadgeView.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using MonoTouch.CoreGraphics;
using MonoTouch.UIKit;

namespace XamarinStore
{
	class BadgeView : UILabel
	{
		const float height = 14;

		public int BadgeNumber {
			get { return badgeNumber; }
			set {
				Text = (badgeNumber = value).ToString ();
				CalculateSize ();
				Alpha = badgeNumber > 0 ? 1 : 0;
				SetNeedsDisplay ();
			}
		}

		SizeF numberSize;
		int badgeNumber;

		public BadgeView ()
		{
			BackgroundColor = UIColor.Clear;
			TextColor = Color.Blue;
			Font = UIFont.BoldSystemFontOfSize (10f);;
			UserInteractionEnabled = false;
			Layer.CornerRadius = height / 2;
			Layer.BackgroundColor = UIColor.White.CGColor;
			TextAlignment = UITextAlignment.Center;
		}

		void CalculateSize ()
		{
			numberSize = StringSize (badgeNumber.ToString (), Font);
			Frame = new RectangleF (Frame.Location, new SizeF (Math.Max (numberSize.Width, height), height));
		}
	}
}

File: F#/XamarinStore.iOS/Views/BadgeView.fs
namespace XamarinStore

open System
open System.Collections.Generic
open System.Drawing
open System.Linq
open System.Text
open MonoTouch.CoreGraphics
open MonoTouch.UIKit
open MonoTouch.Foundation

type BadgeView() as this =
    inherit UILabel()

    let height = 14.0f
    let mutable badgeNumber = 0
    let mutable numberSize = new SizeF()
    do this.BackgroundColor <- UIColor.Clear
       this.TextColor <- Color.Blue.ToUIColor()
       this.Font <- UIFont.BoldSystemFontOfSize 10.0f
       this.UserInteractionEnabled <- false
       this.Layer.CornerRadius <- height / 2.0f
       this.BackgroundColor <- UIColor.White
       this.TextAlignment <- UITextAlignment.Center

    let calculateSize () =
        numberSize <- this.StringSize(badgeNumber.ToString(), this.Font)
        this.Frame <- new RectangleF (this.Frame.Location, new SizeF (Math.Max (numberSize.Width, height), height))

    member this.BadgeNumber
        with get () = badgeNumber
        and set value = badgeNumber <- value
                        this.Text <- value.ToString ()
                        calculateSize ()
                        this.Alpha <- if badgeNumber > 0 then 1.0f else 0.0f
                        this.SetNeedsDisplay ()
